auto-instrumentation/httpurlconnection/library/src/main/java/io/opentelemetry/instrumentation/library/httpurlconnection/HttpUrlReplacements.java

```diff
@@ -0,0 +1,466 @@
+        markHarvestable(c);
+
+        return responseCode;
+    }
+
+    public static synchronized String replacementForResponseMessage(URLConnection c)
+            throws IOException {
+        startTracingAtFirstConnection(c);
+
+        String responseMessage;
+        HttpURLConnection httpURLConnection = (HttpURLConnection) c;
+        try {
+            responseMessage = httpURLConnection.getResponseMessage();
+        } catch (IOException e) {
+            reportWithThrowable(c, e);
+            throw e;
+        }
+
+        updateLastSeenTime(c);
+        markHarvestable(c);
+
+        return responseMessage;
+    }
+
+    public static synchronized OutputStream replacementForOutputStream(URLConnection c)
+            throws IOException {
+        startTracingAtFirstConnection(c);
+
+        OutputStream outputStream;
+        try {
+            outputStream = c.getOutputStream();
+        } catch (IOException e) {
+            reportWithThrowable(c, e);
+            throw e;
+        }
+
+        updateLastSeenTime(c);
+        // getOutputStream() does not read anything from connection so request not harvestable yet
+        // (not reportable if left idle).
+
+        return outputStream;
+    }
+
+    public static synchronized InputStream replacementForInputStream(URLConnection c)
+            throws IOException {
+        startTracingAtFirstConnection(c);
+
+        InputStream inputStream;
+        try {
+            inputStream = c.getInputStream();
+        } catch (IOException e) {
+            reportWithThrowable(c, e);
+            throw e;
+        }
+
+        HttpURLConnection httpURLConnection = (HttpURLConnection) c;
+        reportWithResponseCode(httpURLConnection);
+
+        return inputStream;
+    }
+
+    public static synchronized InputStream replacementForErrorStream(HttpURLConnection c) {
+        startTracingAtFirstConnection(c);
+
+        InputStream errorStream = c.getErrorStream();
+
+        reportWithResponseCode(c);
+
+        return errorStream;
+    }
+
+    private static synchronized void reportWithThrowable(URLConnection c, IOException e) {
+        endTracing(c, UNKNOWN_RESPONSE_CODE, e);
+    }
+
+    private static synchronized void reportWithResponseCode(HttpURLConnection c) {
+        try {
+            endTracing(c, c.getResponseCode(), null);
+        } catch (IOException e) {
+            // TODO: Log instrumentation error in getting response code
+        }
+    }
+
+    private static synchronized void endTracing(
+            URLConnection c, int responseCode, Throwable error) {
+        HttpURLConnectionInfo info = activeURLConnections.get(c);
+        if (info != null && !info.reported) {
+            Context context = info.context;
+            instrumenter().end(context, c, responseCode, error);
+            info.reported = true;
+        }
+    }
+
+    @SuppressWarnings("MustBeClosedChecker")
+    private static synchronized void startTracingAtFirstConnection(URLConnection c) {
```

Comment:
Do we _need_ `synchronized` on the private methods of this class? I think the entire public surface is already covered with `synchronized`, so I think an extra lock step is unnecessary?

Reply:
I looked into it more. The functions which read/write the weak hash map need to be synchronized. We can instead remove synchronization from all the public replacement apis. I will do that change, test and submit the commit.